apis/oam/v1alpha2/core_types.go

```diff
@@ -0,0 +1,375 @@
+/*
+Copyright 2020 The Crossplane Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package v1alpha2
+
+import (
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apimachinery/pkg/types"
+	"k8s.io/apimachinery/pkg/util/intstr"
+
+	runtimev1alpha1 "github.com/crossplane/crossplane-runtime/apis/core/v1alpha1"
+)
+
+// A DefinitionReference refers to a CustomResourceDefinition by name.
+type DefinitionReference struct {
+	// Name of the referenced CustomResourceDefinition.
+	Name string `json:"name"`
+}
+
+// A WorkloadDefinitionSpec defines the desired state of a WorkloadDefinition.
+type WorkloadDefinitionSpec struct {
+	// Reference to the CustomResourceDefinition that defines this workload
+	// kind.
+	Reference DefinitionReference `json:"definitionRef"`
+}
+
+// +kubebuilder:object:root=true
+
+// A WorkloadDefinition registers a kind of Kubernetes custom resource as a
+// valid OAM workload kind by referencing its CustomResourceDefinition. The CRD
+// is used to validate the schema of the workload when it is embedded in an OAM
+// Component.
+// +kubebuilder:printcolumn:JSONPath=".spec.definitionRef.name",name=DEFINITION-NAME,type=string
+type WorkloadDefinition struct {
+	metav1.TypeMeta   `json:",inline"`
+	metav1.ObjectMeta `json:"metadata,omitempty"`
+
+	Spec WorkloadDefinitionSpec `json:"spec,omitempty"`
+}
+
+// +kubebuilder:object:root=true
+
+// WorkloadDefinitionList contains a list of WorkloadDefinition.
+type WorkloadDefinitionList struct {
+	metav1.TypeMeta `json:",inline"`
+	metav1.ListMeta `json:"metadata,omitempty"`
+	Items           []WorkloadDefinition `json:"items"`
+}
+
+// A TraitDefinitionSpec defines the desired state of a TraitDefinition.
+type TraitDefinitionSpec struct {
+	// Reference to the CustomResourceDefinition that defines this trait kind.
+	Reference DefinitionReference `json:"definitionRef"`
+
+	// AppliesToWorkloads specifies the list of workload kinds this trait
+	// applies to. Workload kinds are specified in kind.group/version format,
+	// e.g. server.core.oam.dev/v1alpha2. Traits that omit this field apply to
+	// all workload kinds.
+	// +optional
+	AppliesToWorkloads []string `json:"appliesToWorkloads,omitempty"`
```

Comment:
Since we plan to allow trait to apply to "policy" later, should this still be called `appliesTo` as in alpha1?

Reply:
I don't feel super strongly, but I like that `appliesToWorkloads` is a little more explicit in the current implementation. I don't think what we name this field now really has to affect what we name any fields that configure policy around trait-to-workload compatibility.

Reply:
In the future,  I envision a trait "appliesTo" a policy. Does this not conflict with this name?

Reply:
I don't think so. The policies still describe workloads, right? The trait isn't applying to a policy, it's using policies to determine which workload it applies to.

Regardless of what we name this field now it will be a breaking API change (presuming OAM mirrors Kubernetes API versioning policy) to switch its meaning from a list of workload kinds to a list of policies. It would also be a breaking API change to remove this field and replace its functionality with a differently named field. Fortunately (again presuming we mirror Kubernetes policy) alpha APIs are allowed to make breaking changes without warning.